l2geth/rollup/sync_service.go

```diff
@@ -882,46 +883,51 @@ func (s *SyncService) applyTransactionToTip(tx *types.Transaction) error {
 		s.SetLatestEnqueueIndex(tx.GetMeta().QueueIndex)
 	}
 
-	// The index was set above so it is safe to dereference
-	log.Debug("Applying transaction to tip", "index", *tx.GetMeta().Index, "hash", tx.Hash().Hex(), "origin", tx.QueueOrigin().String())
+	err := func() error {
+		// The index was set above so it is safe to dereference
+		log.Debug("Applying transaction to tip", "index", *tx.GetMeta().Index, "hash", tx.Hash().Hex(), "origin", tx.QueueOrigin().String())
 
-	// Log transaction to the failover log
-	if err := s.publishTransaction(tx); err != nil {
-		log.Error("Failed to publish transaction to log", "msg", err)
-		return fmt.Errorf("internal error: transaction logging failed")
-	}
+		// Log transaction to the failover log
+		if err := s.publishTransaction(tx); err != nil {
+			log.Error("Failed to publish transaction to log", "msg", err)
+			return fmt.Errorf("internal error: transaction logging failed")
+		}
 
-	txs := types.Transactions{tx}
-	errCh := make(chan error, 1)
-	s.txFeed.Send(core.NewTxsEvent{
-		Txs:   txs,
-		ErrCh: errCh,
-	})
-	// Block until the transaction has been added to the chain
-	log.Trace("Waiting for transaction to be added to chain", "hash", tx.Hash().Hex())
+		txs := types.Transactions{tx}
+		errCh := make(chan error, 1)
+		s.txFeed.Send(core.NewTxsEvent{
+			Txs:   txs,
+			ErrCh: errCh,
+		})
+		// Block until the transaction has been added to the chain
+		log.Trace("Waiting for transaction to be added to chain", "hash", tx.Hash().Hex())
+
+		select {
+		case err := <-errCh:
+			log.Error("Got error waiting for transaction to be added to chain", "msg", err)
+			return err
+		case <-s.chainHeadCh:
+			// Update the cache when the transaction is from the owner
+			// of the gas price oracle
+			sender, _ := types.Sender(s.signer, tx)
+			owner := s.GasPriceOracleOwnerAddress()
+			if owner != nil && sender == *owner {
+				if err := s.updateGasPriceOracleCache(nil); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+	}()
 
-	select {
-	case err := <-errCh:
-		log.Error("Got error waiting for transaction to be added to chain", "msg", err)
+	// Rollback state on errors
+	if err != nil {
 		s.SetLatestL1Timestamp(ts)
 		s.SetLatestL1BlockNumber(bn)
 		s.SetLatestIndex(index)
-		return err
-	case <-s.chainHeadCh:
-		// Update the cache when the transaction is from the owner
-		// of the gas price oracle
-		sender, _ := types.Sender(s.signer, tx)
-		owner := s.GasPriceOracleOwnerAddress()
-		if owner != nil && sender == *owner {
-			if err := s.updateGasPriceOracleCache(nil); err != nil {
-				s.SetLatestL1Timestamp(ts)
-				s.SetLatestL1BlockNumber(bn)
-				s.SetLatestIndex(index)
-				return err
-			}
-		}
-		return nil
+		s.SetLatestEnqueueIndex(queueIndex)
```

Comment:
when the deadlock bug was fixed, there was a reason we decided to leave out resetting the enqueue index. i can't recall what the exact reason was, but wanted to bring it up so that we can revisit those assumptions